src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -358,6 +358,109 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
+            for (var i = 0; i < genericArgumentDefinitions.Length; ++i)
+            {
+                var argumentDefinitionTypeInfo = genericArgumentDefinitions[i].GetTypeInfo();
+                var parameter = parameters[i];
+                var parameterTypeInfo = parameter.GetTypeInfo();
+                foreach (var constraint in argumentDefinitionTypeInfo.GetGenericParameterConstraints())
+                {
+                    var substituteGenericParameterConstraint = SubstituteGenericParameterConstraint(parameters, constraint);
+                    if (!ParameterCompatibleWithTypeConstraint(parameter, substituteGenericParameterConstraint))
+                    {
+                        return false;
+                    }
+                }
+
+                var specialConstraints = argumentDefinitionTypeInfo.GenericParameterAttributes;
+                if ((specialConstraints & GenericParameterAttributes.DefaultConstructorConstraint) == GenericParameterAttributes.DefaultConstructorConstraint)
+                {
+                    if (!parameterTypeInfo.IsValueType && parameterTypeInfo.DeclaredConstructors.Where(c => c.IsPublic).All(c => c.GetParameters().Length != 0))
+                    {
+                        return false;
+                    }
+                }
+                if ((specialConstraints & GenericParameterAttributes.ReferenceTypeConstraint) == GenericParameterAttributes.ReferenceTypeConstraint)
+                {
+                    if (parameterTypeInfo.IsValueType)
+                    {
+                        return false;
+                    }
+                }
+                if ((specialConstraints & GenericParameterAttributes.NotNullableValueTypeConstraint) == GenericParameterAttributes.NotNullableValueTypeConstraint)
+                {
+                    if (!parameterTypeInfo.IsValueType || parameterTypeInfo.IsGenericType && IsGenericTypeDefinedBy(parameter, typeof(Nullable<>)))
+                    {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+
+        private static bool IsGenericTypeDefinedBy(Type type, Type openGeneric)
+        {
+            return !type.ContainsGenericParameters && type.IsGenericType && type.GetGenericTypeDefinition() == openGeneric;
+        }
+
+        private static Type SubstituteGenericParameterConstraint(Type[] parameters, Type constraint)
+        {
+            if (!constraint.IsGenericParameter) return constraint;
+            return parameters[constraint.GenericParameterPosition];
+        }
+
+        private static bool ParameterCompatibleWithTypeConstraint(Type parameter, Type constraint)
+        {
+            bool hasAnyCompatibleType = false;
+            foreach (var p in new[] {parameter, parameter.BaseType}.Concat(parameter.GetTypeInfo().ImplementedInterfaces))
```

Comment:
We avoid LINQ because of the number of allocations it produces. This method allocates way too much for what it does.